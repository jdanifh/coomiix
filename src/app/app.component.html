<cmx-layout [title]="title">
  <cmx-locale header></cmx-locale>
  <cmx-side-menu menu></cmx-side-menu>
  <router-outlet content></router-outlet>
</cmx-layout>
