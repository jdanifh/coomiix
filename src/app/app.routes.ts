import { Routes } from '@angular/router';

export const routes: Routes = [
  {
    path: '**',
    title: 'Home',
    loadComponent: () => import('./views/home/home.component').then(mod => mod.HomeComponent)
  },
];
