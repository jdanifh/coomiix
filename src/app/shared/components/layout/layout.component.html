<header [class.closed-menu]="closed()">
  <a routerLink="/" class="title" matRipple>
    <span>{{ title() }}</span>
  </a>
  <button (click)="toggleMenu()" mat-icon-button aria-label="toggle menu">
    <mat-icon aria-hidden="true">more_vert</mat-icon>
  </button>
  <!-- header slot -->
  <ng-content select="[header]"></ng-content>
</header>
<aside [class.closed-menu]="closed()">
  <!-- menu slot -->
  <ng-content select="[menu]">
    <a routerLink="/" routerLinkActive="active" matRipple>
      <mat-icon aria-hidden="true">home</mat-icon>
      <span>Home</span>
    </a>
  </ng-content>
</aside>
<main>
  <!-- content slot -->
  <ng-content select="[content]"></ng-content>
</main>
