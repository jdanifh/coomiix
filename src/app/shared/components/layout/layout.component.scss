:host {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

header {
  height: 100px;
  display: flex;
  flex: 1 1 100%;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  background-color: var(--background-color);
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.5);

  .title {
    text-decoration: none;
    color: var(--primary-text-color);
    font-size: 2em;
    font-weight: 700;
    padding: 25px;
  }

  @media (width < 600px) {
    .title + button {
      margin-left: auto;
      margin-right: 25px;
    }

    .content {
      flex: 1 1 100%;
    }
  }

  & .title + button {
    transition: all 0.5s ease;
    transform: rotate(270deg);
  }
  &.closed-menu .title + button {
    transform: rotate(0deg);
  }
}

aside {
  width: 200px;
  padding: 25px;
  transition: all 0.5s ease;

  ::ng-deep a {
    padding: 25px;
    color: var(--primary-text-color);
    background-color: var(--primary-light-color);
    text-decoration: none;
    display: flex;
    border-radius: 16px;
    align-items: center;
    justify-content: center;

    &:hover, &.active {
      background-color: var(--primary-color);
    }

    & > mat-icon {
      margin: 0px 20px 0px -20px;
    }

    & + a {
      margin-top: 25px;
    }
  }

  &.closed-menu {
    width: 80px;

    ::ng-deep a {

      & > mat-icon {
        margin: 0px;
      }

      & > span {
        display: none;
      }
    }
  }


  @media (width < 600px) {
    position: fixed;
    top: 0;
    left: 0;
    z-index: 1000;
    width: 90vw;
    height: calc(100vh - 100px);
    margin-top: 100px;

    &.closed-menu {
      left: 100vw;
    }
  }
}

main {
  padding: 50px;
  flex: 1;
  transition: all 0.5s ease;
}
