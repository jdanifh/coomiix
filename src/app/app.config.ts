import { ApplicationConfig, provideExperimentalZonelessChangeDetection, provideZoneChangeDetection } from '@angular/core';
import { provideAnimationsAsync } from '@angular/platform-browser/animations/async';
import { provideRouter, withDebugTracing, withInMemoryScrolling, withViewTransitions } from '@angular/router';
import { provideHttpClient, withInterceptors } from '@angular/common/http';
import { MAT_FORM_FIELD_DEFAULT_OPTIONS } from '@angular/material/form-field';
import { provideLocales } from './core/providers/locale/locale.provider';

import { routes } from './app.routes';

// Zone.js is disabled in favor of zoneless change detection
export const appConfig: ApplicationConfig = {
  providers: [
    {provide: MAT_FORM_FIELD_DEFAULT_OPTIONS, useValue: {appearance: 'outline', floatLabel: 'always'}},
    provideExperimentalZonelessChangeDetection(),
    provideAnimationsAsync(),
    provideRouter(routes, withDebugTracing(), withViewTransitions(), withInMemoryScrolling()),
    provideHttpClient(withInterceptors([])),
    provideLocales(),
  ]
};
