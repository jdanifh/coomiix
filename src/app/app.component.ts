import { Component, ChangeDetectionStrategy } from '@angular/core';
import { RouterOutlet } from '@angular/router';
import { LayoutComponent } from "./shared/components/layout/layout.component";
import { SideMenuComponent } from "./shared/components/side-menu/side-menu.component";
import { LocaleComponent } from "./shared/components/locale/locale.component";
import { BreadcrumbsComponent } from "./shared/components/breadcrumbs/breadcrumbs.component";

@Component({
  selector: 'app-root',
  standalone: true,
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [
    RouterOutlet,
    LayoutComponent,
    SideMenuComponent,
    LocaleComponent,
    BreadcrumbsComponent
],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss'
})
export class AppComponent {
  public title = 'Coomiix';
}
