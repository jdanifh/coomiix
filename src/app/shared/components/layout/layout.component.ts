import { ChangeDetectionStrategy, Component, inject, input, signal } from '@angular/core';
import { RouterModule, Router, NavigationEnd } from '@angular/router';
import { MatIconModule } from '@angular/material/icon';
import { MatButtonModule } from '@angular/material/button';
import { BreakpointObserver } from '@angular/cdk/layout';
import { filter } from 'rxjs';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';

@Component({
  selector: 'cmx-layout',
  imports: [
    RouterModule,
    MatIconModule,
    MatButtonModule,
  ],
  templateUrl: './layout.component.html',
  styleUrl: './layout.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class LayoutComponent {

  public title = input('App');
  public closed = signal(false);

  private _observer = inject(BreakpointObserver);
  private _router = inject(Router);

  constructor() {
    if (this._observer.isMatched('(width < 960px)')) {
      this.closed.set(true);
      this._router.events.pipe(
        takeUntilDestroyed(),
        filter(event => event instanceof NavigationEnd)
      ).subscribe(() => this.closed.set(true));
    }
  }

  public toggleMenu() {
    this.closed.update(value => !value);
  }

}
