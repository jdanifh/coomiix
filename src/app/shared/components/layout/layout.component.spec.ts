import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LayoutComponent } from './layout.component';
import { Event, NavigationEnd, provideRouter, Router } from '@angular/router';
import { routes } from '../../../app.routes';
import { By } from '@angular/platform-browser';
import { BreakpointObserver } from '@angular/cdk/layout';
import { EventEmitter } from '@angular/core';

describe('LayoutComponent', () => {
  let component: LayoutComponent;
  let fixture: ComponentFixture<LayoutComponent>;
  let observer: BreakpointObserver;
  let router: Router;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [LayoutComponent],
      providers: [provideRouter(routes)]
    })
    .compileComponents();

    observer = TestBed.inject(BreakpointObserver);
    router = TestBed.inject(Router);
    fixture = TestBed.createComponent(LayoutComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render App title', () => {
    const title = fixture.debugElement.query(By.css('.title'));
    expect(title.nativeElement.textContent).toBe('App');
  });

  it(`should initialize opened menu`, () => {
    expect(component.closed()).toBe(false);
  })

  it(`should toggle menu`, () => {
    expect(component.closed()).toBe(false);
    const menuButton = fixture.debugElement.query(By.css('.title + button'));
    menuButton.triggerEventHandler('click');
    fixture.detectChanges();
    expect(component.closed()).toBe(true);
  })

  it(`should initialize closed menu for small devices`, () => {
    const spy = jest.spyOn(observer, 'isMatched').mockReturnValue(true);
    const smallFixture = TestBed.createComponent(LayoutComponent);
    const smallComponent = smallFixture.componentInstance;
    smallFixture.detectChanges();
    expect(spy).toHaveBeenCalled();
    expect(smallComponent.closed()).toBe(true);
  })

  it(`should close on navigate for small devices`, () => {
    const spy = jest.spyOn(observer, 'isMatched').mockReturnValue(true);
    const events = new EventEmitter<Event>();
    jest.spyOn(router, 'events', 'get').mockReturnValue(events.asObservable());
    const smallFixture = TestBed.createComponent(LayoutComponent);
    const smallComponent = smallFixture.componentInstance;

    smallFixture.detectChanges();
    expect(spy).toHaveBeenCalled();
    expect(smallComponent.closed()).toBe(true);

    smallComponent.toggleMenu();
    smallFixture.detectChanges();
    expect(smallComponent.closed()).toBe(false);

    events.next(new NavigationEnd(1, '/', '/'))
    smallFixture.detectChanges();
    expect(smallComponent.closed()).toBe(true);

  })

  it(`should match snapshot`, () => {
    expect(fixture.nativeElement.innerHTML).toMatchSnapshot();
  })

});
