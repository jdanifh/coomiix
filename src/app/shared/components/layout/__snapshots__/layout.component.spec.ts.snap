// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutComponent should match snapshot 1`] = `"<header><a routerlink="/" matripple="" class="mat-ripple title" ng-reflect-router-link="/" href="/"><span>App</span></a><button mat-icon-button="" aria-label="toggle menu" mat-ripple-loader-uninitialized="" mat-ripple-loader-class-name="mat-mdc-button-ripple" class="mdc-icon-button mat-mdc-icon-button mat-unthemed mat-mdc-button-base" mat-ripple-loader-centered=""><span class="mat-mdc-button-persistent-ripple mdc-icon-button__ripple"></span><mat-icon role="img" aria-hidden="true" class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color" data-mat-icon-type="font">more_vert</mat-icon><span class="mat-focus-indicator"></span><span class="mat-mdc-button-touch-target"></span></button></header><aside><a routerlink="/" routerlinkactive="active" matripple="" class="mat-ripple" ng-reflect-router-link="/" ng-reflect-router-link-active="active" href="/"><mat-icon role="img" aria-hidden="true" class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color" data-mat-icon-type="font">home</mat-icon><span>Home</span></a><!--container--></aside><main></main>"`;
